File: rust/periph/src/chips/pressure/lps28dfw.rs
//! LPS28DFW — dual full-scale digital barometer (STMicroelectronics).
//!
//! Communicates over I²C at address 0x5C (SA0=GND) or 0x5D (SA0=VDD).
//! 24-bit pressure (signed) and 16-bit temperature (signed, 0.01 °C LSB).
//!
//! ## Constants
//!
//! ODR: [`ODR_POWER_DOWN`] through [`ODR_200_HZ`]
//! AVG: [`AVG_4`] through [`AVG_512`]
//! Full-scale: [`FS_MODE_1`], [`FS_MODE_2`]
//! IIR filter: [`LFPF_ODR_OVER_4`], [`LFPF_ODR_OVER_9`]
//! FIFO mode: [`FIFO_BYPASS`] through [`FIFO_CONTINUOUS_TO_FIFO`]
//! Status flags: [`STATUS_P_DA`], [`STATUS_T_DA`], [`STATUS_P_OR`], [`STATUS_T_OR`]

use embedded_hal::i2c::I2c;

const REG_INTERRUPT_CFG: u8 = 0x0B;
const REG_THS_P_L: u8 = 0x0C;
const REG_THS_P_H: u8 = 0x0D;
const REG_WHO_AM_I: u8 = 0x0F;
const REG_CTRL_REG1: u8 = 0x10;
const REG_CTRL_REG2: u8 = 0x11;
const REG_STATUS: u8 = 0x27;
const REG_PRESS_OUT_XL: u8 = 0x28;
const REG_TEMP_OUT_L: u8 = 0x2B;
const REG_RPDS_L: u8 = 0x1A;
const REG_RPDS_H: u8 = 0x1B;
const REG_FIFO_CTRL: u8 = 0x14;
const REG_FIFO_WTM: u8 = 0x15;
const REG_FIFO_STATUS1: u8 = 0x25;
const REG_FIFO_DATA_PRESS_XL: u8 = 0x78;

const CHIP_ID: u8 = 0xB4;
const BOOT_WAIT_MS: u32 = 2;

/// ODR code: power-down / one-shot.
pub const ODR_POWER_DOWN: u8 = 0x00;
/// ODR code: 1 Hz.
pub const ODR_1_HZ: u8 = 0x01;
/// ODR code: 4 Hz.
pub const ODR_4_HZ: u8 = 0x02;
/// ODR code: 10 Hz.
pub const ODR_10_HZ: u8 = 0x03;
/// ODR code: 25 Hz.
pub const ODR_25_HZ: u8 = 0x04;
/// ODR code: 50 Hz.
pub const ODR_50_HZ: u8 = 0x05;
/// ODR code: 75 Hz.
pub const ODR_75_HZ: u8 = 0x06;
/// ODR code: 100 Hz.
pub const ODR_100_HZ: u8 = 0x07;
/// ODR code: 200 Hz.
pub const ODR_200_HZ: u8 = 0x08;

/// Averaging: 4 samples.
pub const AVG_4: u8 = 0x00;
/// Averaging: 8 samples.
pub const AVG_8: u8 = 0x01;
/// Averaging: 16 samples.
pub const AVG_16: u8 = 0x02;
/// Averaging: 32 samples.
pub const AVG_32: u8 = 0x03;
/// Averaging: 64 samples.
pub const AVG_64: u8 = 0x04;
/// Averaging: 128 samples.
pub const AVG_128: u8 = 0x05;
/// Averaging: 512 samples.
pub const AVG_512: u8 = 0x07;

/// Full-scale mode 1 (0–1260 hPa, 4096 LSB/hPa).
pub const FS_MODE_1: u8 = 0;
/// Full-scale mode 2 (0–4060 hPa, 2048 LSB/hPa).
pub const FS_MODE_2: u8 = 1;

/// IIR low-pass filter bandwidth: ODR/4.
pub const LFPF_ODR_OVER_4: u8 = 0;
/// IIR low-pass filter bandwidth: ODR/9.
pub const LFPF_ODR_OVER_9: u8 = 1;

/// FIFO mode: bypass.
pub const FIFO_BYPASS: u8 = 0;
/// FIFO mode: FIFO.
pub const FIFO_FIFO: u8 = 1;
/// FIFO mode: continuous.
pub const FIFO_CONTINUOUS: u8 = 2;
/// FIFO mode: bypass-to-FIFO.
pub const FIFO_BYPASS_TO_FIFO: u8 = 4;
/// FIFO mode: bypass-to-continuous.
pub const FIFO_BYPASS_TO_CONTINUOUS: u8 = 5;
/// FIFO mode: continuous-to-FIFO.
pub const FIFO_CONTINUOUS_TO_FIFO: u8 = 6;

/// STATUS flag: new pressure data available.
pub const STATUS_P_DA: u8 = 0x01;
/// STATUS flag: new temperature data available.
pub const STATUS_T_DA: u8 = 0x02;
/// STATUS flag: pressure overrun.
pub const STATUS_P_OR: u8 = 0x10;
/// STATUS flag: temperature overrun.
pub const STATUS_T_OR: u8 = 0x20;

fn write_reg<I2C: I2c>(i2c: &mut I2C, addr: u8, reg: u8, value: u8) -> Result<(), I2C::Error> {
    i2c.write(addr, &[reg, value])
}

fn read_reg_bytes<I2C: I2c>(i2c: &mut I2C, addr: u8, reg: u8, buf: &mut [u8]) -> Result<(), I2C::Error> {
    i2c.write_read(addr, &[reg], buf)
}

/// LPS28DFW minimal driver — pressure (hPa) and temperature (°C).
///
/// Default configuration: Mode 1, AVG=16, ODR=25 Hz, BDU=1, IIR filter enabled.
pub struct Lps28dfwMinimal<I2C> {
    i2c: I2C,
    addr: u8,
    fs_mode: u8,
    odr: u8,
    avg: u8,
    lpf_en: u8,
    lpf_cfg: u8,
    bdu: u8,
}

impl<I2C: I2c> Lps28dfwMinimal<I2C> {
    /// Create a new `Lps28dfwMinimal`.
    ///
    /// # Arguments
    /// * `i2c` — Configured I²C bus.
    /// * `addr` — 7-bit I²C address (0x5C or 0x5D).
    pub fn new(mut i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        let mut s = Self {
            i2c, addr,
            fs_mode: 0, odr: 0x04, avg: 0x02,
            lpf_en: 1, lpf_cfg: 0, bdu: 1,
        };
        let ctrl2 = (s.fs_mode << 6) | (s.lpf_cfg << 5) | (s.lpf_en << 4) | (s.bdu << 3);
        write_reg(&mut s.i2c, s.addr, REG_CTRL_REG2, ctrl2)?;
        let ctrl1 = (s.odr << 3) | (s.avg & 0x07);
        write_reg(&mut s.i2c, s.addr, REG_CTRL_REG1, ctrl1)?;
        Ok(s)
    }

    fn read_pressure_raw(&mut self) -> Result<i32, I2C::Error> {
        let mut buf = [0u8; 3];
        read_reg_bytes(&mut self.i2c, self.addr, REG_PRESS_OUT_XL, &mut buf)?;
        let v = ((buf[2] as i32) << 16) | ((buf[1] as i32) << 8) | (buf[0] as i32);
        Ok(if v & 0x800000 != 0 { v | -0x1000000 } else { v })
    }

    fn read_temperature_raw(&mut self) -> Result<i16, I2C::Error> {
        let mut buf = [0u8; 2];
        read_reg_bytes(&mut self.i2c, self.addr, REG_TEMP_OUT_L, &mut buf)?;
        Ok(i16::from_be_bytes([buf[0], buf[1]]))
    }

    /// Read absolute pressure in hPa.
    pub fn read_pressure(&mut self) -> Result<f32, I2C::Error> {
        let raw = self.read_pressure_raw()?;
        let sens = if self.fs_mode == 0 { 4096.0_f32 } else { 2048.0_f32 };
        Ok(raw as f32 / sens)
    }

    /// Read temperature in °C.
    pub fn read_temperature(&mut self) -> Result<f32, I2C::Error> {
        let raw = self.read_temperature_raw()?;
        Ok(raw as f32 / 100.0)
    }
}

/// LPS28DFW full driver — extends minimal with full configuration, FIFO, threshold, etc.
pub struct Lps28dfwFull<I2C> {
    inner: Lps28dfwMinimal<I2C>,
}

impl<I2C: I2c> Lps28dfwFull<I2C> {
    /// Create a new `Lps28dfwFull`.
    pub fn new(i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        Ok(Self { inner: Lps28dfwMinimal::new(i2c, addr)? })
    }

    /// Set output data rate, averaging, full-scale mode, and IIR filter.
    pub fn configure(&mut self, odr: u8, avg: u8, fs_mode: u8, lpf_en: bool, lpf_cfg: u8) -> Result<(), I2C::Error> {
        self.inner.odr = odr;
        self.inner.avg = avg;
        self.inner.fs_mode = fs_mode;
        self.inner.lpf_en = if lpf_en { 1 } else { 0 };
        self.inner.lpf_cfg = lpf_cfg;
        let ctrl2 = (self.inner.fs_mode << 6) | (self.inner.lpf_cfg << 5) | (self.inner.lpf_en << 4) | (self.inner.bdu << 3);
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG2, ctrl2)?;
        let ctrl1 = (self.inner.odr << 3) | (self.inner.avg & 0x07);
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG1, ctrl1)
    }

    /// Burst-read pressure and temperature.
    pub fn read(&mut self) -> Result<(f32, f32), I2C::Error> {
        let mut buf = [0u8; 5];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_PRESS_OUT_XL, &mut buf)?;
        let p = ((buf[2] as i32) << 16) | ((buf[1] as i32) << 8) | (buf[0] as i32);
        let p = if p & 0x800000 != 0 { p | -0x1000000 } else { p };
        let t = i16::from_be_bytes([buf[3], buf[4]]);
        let sens = if self.inner.fs_mode == 0 { 4096.0_f32 } else { 2048.0_f32 };
        Ok((p as f32 / sens, t as f32 / 100.0))
    }

    /// Check whether STATUS.P_DA is set.
    pub fn is_data_ready(&mut self) -> Result<bool, I2C::Error> {
        let mut buf = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_STATUS, &mut buf)?;
        Ok(buf[0] & STATUS_P_DA != 0)
    }

    /// Trigger a one-shot measurement (with ODR=0000) and read the result.
    pub fn read_oneshot(&mut self) -> Result<(f32, f32), I2C::Error> {
        let mut saved = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG1, &mut saved)?;
        let saved_odr = saved[0] >> 3;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG1, self.inner.avg & 0x07)?;
        let mut c2 = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG2, &mut c2)?;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG2, c2[0] | 0x01)?;
        let result = self.read();
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG1, (saved_odr << 3) | (self.inner.avg & 0x07))?;
        result
    }

    /// Program the one-point calibration offset (RPDS).
    pub fn set_offset(&mut self, offset_hpa: f32) -> Result<(), I2C::Error> {
        let sens = if self.inner.fs_mode == 0 { 4096.0_f32 } else { 2048.0_f32 };
        let mut raw = (offset_hpa * sens) as i32;
        if raw < 0 { raw += 0x10000; }
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_RPDS_L, (raw & 0xFF) as u8)?;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_RPDS_H, ((raw >> 8) & 0xFF) as u8)
    }

    /// Issue a software reset and wait for the chip to reboot.
    pub fn softreset(&mut self) -> Result<(), I2C::Error> {
        let mut c2 = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG2, &mut c2)?;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_CTRL_REG2, c2[0] | 0x02)
    }

    /// Configure FIFO mode, watermark level, and stop-on-watermark.
    pub fn fifo_configure(&mut self, mode: u8, wtm: u8, stop_on_wtm: bool) -> Result<(), I2C::Error> {
        if mode == FIFO_BYPASS {
            write_reg(&mut self.inner.i2c, self.inner.addr, REG_FIFO_CTRL, 0x00)?;
        }
        let trig = if mode >= 4 { 1 } else { 0 };
        let f_mode = mode & 0x03;
        let ctrl = (trig << 2) | ((if stop_on_wtm { 1 } else { 0 }) << 3) | f_mode;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_FIFO_CTRL, ctrl)?;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_FIFO_WTM, wtm & 0x7F)
    }

    /// Drain up to `count` pressure samples from the FIFO.
    pub fn fifo_read(&mut self, count: u8, buf: &mut [f32]) -> Result<(), I2C::Error> {
        let n = if count as usize > buf.len() { buf.len() as u8 } else { count };
        if n == 0 { return Ok(()); }
        let mut raw = [0u8; 384];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_FIFO_DATA_PRESS_XL, &mut raw[..(n as usize) * 3])?;
        let sens = if self.inner.fs_mode == 0 { 4096.0_f32 } else { 2048.0_f32 };
        for i in 0..n as usize {
            let b = i * 3;
            let v = ((raw[b + 2] as i32) << 16) | ((raw[b + 1] as i32) << 8) | (raw[b] as i32);
            let v = if v & 0x800000 != 0 { v | -0x1000000 } else { v };
            buf[i] = v as f32 / sens;
        }
        Ok(())
    }

    /// Return the number of unread samples in the FIFO.
    pub fn fifo_level(&mut self) -> Result<u8, I2C::Error> {
        let mut buf = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_FIFO_STATUS1, &mut buf)?;
        Ok(buf[0])
    }

    /// Program the pressure threshold and enable interrupt sources.
    pub fn set_threshold(&mut self, threshold_hpa: f32, high: bool, low: bool) -> Result<(), I2C::Error> {
        let sens = if self.inner.fs_mode == 0 { 16.0_f32 } else { 8.0_f32 };
        let mut raw = (threshold_hpa * sens) as i32;
        if raw < 0 { raw = 0; }
        if raw > 0x7FFF { raw = 0x7FFF; }
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_THS_P_L, (raw & 0xFF) as u8)?;
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_THS_P_H, ((raw >> 8) & 0x7F) as u8)?;
        let mut cfg = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_INTERRUPT_CFG, &mut cfg)?;
        let mut v = cfg[0] & !0x03;
        if high { v |= 0x01; }
        if low  { v |= 0x02; }
        write_reg(&mut self.inner.i2c, self.inner.addr, REG_INTERRUPT_CFG, v)
    }

    /// Read the WHO_AM_I register.
    pub fn chip_id(&mut self) -> Result<u8, I2C::Error> {
        let mut buf = [0u8; 1];
        read_reg_bytes(&mut self.inner.i2c, self.inner.addr, REG_WHO_AM_I, &mut buf)?;
        Ok(buf[0])
    }

    /// Compute altitude above sea level from the current pressure.
    pub fn altitude(&mut self, sea_level_hpa: f32) -> Result<f32, I2C::Error> {
        let p = self.inner.read_pressure()?;
        Ok(44330.0 * (1.0 - libm::powf(p / sea_level_hpa, 1.0 / 5.255)))
    }

    /// Read calibrated pressure (delegated from Minimal).
    pub fn read_pressure(&mut self) -> Result<f32, I2C::Error> { self.inner.read_pressure() }

    /// Read calibrated temperature (delegated from Minimal).
    pub fn read_temperature(&mut self) -> Result<f32, I2C::Error> { self.inner.read_temperature() }
}

File: python/periph/chips/pressure/lps28dfw.py
import struct
import time


class LPS28DFWMinimal:
    """LPS28DFW dual full-scale digital barometer — minimal interface.

    Reads absolute pressure and temperature from the CCLGA-7L water-resistant
    sensor. I²C address is 0x5C (SA0=GND) or 0x5D (SA0=VDD). Pressure and
    temperature are returned in SI units (hPa and °C).

    Default configuration (baked in at construction):
        - FS_MODE = 0 (Mode 1, 0–1260 hPa, 4096 LSB/hPa)
        - AVG = 010b (16 samples)
        - ODR = 0100b (25 Hz)
        - BDU = 1, EN_LPFP = 1, LFPF_CFG = 0 (ODR/4 bandwidth)

    Args:
        connection: Configured I²C connection pointing at the device.
    """

    _REG_INTERRUPT_CFG = 0x0B
    _REG_WHO_AM_I      = 0x0F
    _REG_CTRL_REG1     = 0x10
    _REG_CTRL_REG2     = 0x11
    _REG_CTRL_REG3     = 0x12
    _REG_STATUS        = 0x27
    _REG_PRESS_OUT_XL  = 0x28
    _REG_PRESS_OUT_L   = 0x29
    _REG_PRESS_OUT_H   = 0x2A
    _REG_TEMP_OUT_L    = 0x2B
    _REG_TEMP_OUT_H    = 0x2C

    _CHIP_ID        = 0xB4
    _SENSITIVITY_LSB_PER_HPA_MODE1 = 4096.0
    _SENSITIVITY_LSB_PER_HPA_MODE2 = 2048.0
    _BOOT_WAIT_MS   = 2

    def __init__(self, connection):
        self._connection = connection
        self._fs_mode = 0
        self._odr = 0x04
        self._avg = 0x02
        self._lpf_en = 1
        self._lpf_cfg = 0
        self._bdu = 1
        self._init()

    def _init(self):
        who = self._read_reg(self._REG_WHO_AM_I, 1)[0]
        if who != self._CHIP_ID:
            raise OSError('LPS28DFW WHO_AM_I mismatch: expected 0x{:02X}, got 0x{:02X}'.format(self._CHIP_ID, who))
        time.sleep(self._BOOT_WAIT_MS / 1000.0)
        ctrl2 = (self._fs_mode << 6) | (self._lpf_cfg << 5) | (self._lpf_en << 4) | (self._bdu << 3)
        self._write_reg(self._REG_CTRL_REG2, ctrl2)
        ctrl1 = (self._odr << 3) | (self._avg & 0x07)
        self._write_reg(self._REG_CTRL_REG1, ctrl1)

    def _write_reg(self, reg, value):
        self._connection.write(bytes([reg, value]))

    def _read_reg(self, reg, n):
        return self._connection.write_read(bytes([reg]), n)

    def _read_pressure_raw(self):
        raw = self._read_reg(self._REG_PRESS_OUT_XL, 3)
        value = (raw[2] << 16) | (raw[1] << 8) | raw[0]
        if value & 0x800000:
            value -= 0x1000000
        return value

    def _read_temperature_raw(self):
        raw = self._read_reg(self._REG_TEMP_OUT_L, 2)
        value = (raw[1] << 8) | raw[0]
        if value & 0x8000:
            value -= 0x10000
        return value

    def read_pressure(self):
        """Read absolute pressure.

        Returns:
            float: Pressure in hPa. Sensitivity is 4096 LSB/hPa in Mode 1
            (default) and 2048 LSB/hPa in Mode 2 (260–4060 hPa range).
        """
        raw = self._read_pressure_raw()
        if self._fs_mode == 0:
            return raw / self._SENSITIVITY_LSB_PER_HPA_MODE1
        return raw / self._SENSITIVITY_LSB_PER_HPA_MODE2

    def read_temperature(self):
        """Read sensor temperature.

        Returns:
            float: Temperature in degrees Celsius (0.01 °C LSB).
        """
        raw = self._read_temperature_raw()
        return raw / 100.0


class LPS28DFWFull(LPS28DFWMinimal):
    """LPS28DFW full interface — extends LPS28DFWMinimal with full configuration.

    Adds full-scale selection, ODR / averaging control, IIR low-pass filter,
    block-data-update, one-shot trigger, one-point calibration offset, FIFO
    configuration / drain / level, and pressure-threshold interrupt setup.

    Args:
        connection: Configured I²C connection pointing at the device.
    """

    ODR_POWER_DOWN = 0x00
    ODR_1_HZ       = 0x01
    ODR_4_HZ       = 0x02
    ODR_10_HZ      = 0x03
    ODR_25_HZ      = 0x04
    ODR_50_HZ      = 0x05
    ODR_75_HZ      = 0x06
    ODR_100_HZ     = 0x07
    ODR_200_HZ     = 0x08

    AVG_4   = 0x00
    AVG_8   = 0x01
    AVG_16  = 0x02
    AVG_32  = 0x03
    AVG_64  = 0x04
    AVG_128 = 0x05
    AVG_512 = 0x07

    FS_MODE_1 = 0
    FS_MODE_2 = 1

    LFPF_ODR_OVER_4 = 0
    LFPF_ODR_OVER_9 = 1

    FIFO_BYPASS       = 0
    FIFO_FIFO         = 1
    FIFO_CONTINUOUS   = 2
    FIFO_BYPASS_TO_FIFO       = 4
    FIFO_BYPASS_TO_CONTINUOUS = 5
    FIFO_CONTINUOUS_TO_FIFO   = 6

    STATUS_P_DA = 0x01
    STATUS_T_DA = 0x02
    STATUS_P_OR = 0x10
    STATUS_T_OR = 0x20

    INT_SRC_BOOT = 0x80
    INT_SRC_IA   = 0x04
    INT_SRC_PL   = 0x02
    INT_SRC_PH   = 0x01

    def __init__(self, connection):
        super().__init__(connection)

    def configure(self, odr=0x04, avg=0x02, fs_mode=0, lpf_en=True, lpf_cfg=0):
        """Set output data rate, averaging, full-scale mode, and IIR filter.

        Args:
            odr: Output data rate code (0=power-down, 1–7=1–100 Hz, 8=200 Hz).
            avg: Averaging code (0–5 for 4/8/16/32/64/128 samples, 7=512).
            fs_mode: 0=Mode 1 (0–1260 hPa), 1=Mode 2 (0–4060 hPa).
            lpf_en: True to enable the IIR low-pass filter on pressure output.
            lpf_cfg: 0=ODR/4 bandwidth, 1=ODR/9 bandwidth.
        """
        self._odr = odr
        self._avg = avg
        self._fs_mode = fs_mode
        self._lpf_en = 1 if lpf_en else 0
        self._lpf_cfg = lpf_cfg
        ctrl2 = (self._fs_mode << 6) | (self._lpf_cfg << 5) | (self._lpf_en << 4) | (self._bdu << 3)
        self._write_reg(self._REG_CTRL_REG2, ctrl2)
        ctrl1 = (self._odr << 3) | (self._avg & 0x07)
        self._write_reg(self._REG_CTRL_REG1, ctrl1)

    def read(self):
        """Burst-read pressure and temperature.

        Returns:
            dict: ``{"pressure": float hPa, "temperature": float °C}``.
        """
        raw = self._read_reg(self._REG_PRESS_OUT_XL, 5)
        p = (raw[2] << 16) | (raw[1] << 8) | raw[0]
        if p & 0x800000:
            p -= 0x1000000
        t = (raw[4] << 8) | raw[3]
        if t & 0x8000:
            t -= 0x10000
        if self._fs_mode == 0:
            pressure = p / self._SENSITIVITY_LSB_PER_HPA_MODE1
        else:
            pressure = p / self._SENSITIVITY_LSB_PER_HPA_MODE2
        return {"pressure": pressure, "temperature": t / 100.0}

    def read_oneshot(self):
        """Trigger a one-shot measurement (with ODR=0000) and read the result.

        Returns:
            dict: ``{"pressure": float hPa, "temperature": float °C}``.

        Raises:
            OSError: If P_DA does not assert within ~1 s.
        """
        ctrl1 = self._read_reg(self._REG_CTRL_REG1, 1)[0]
        saved_odr = ctrl1 >> 3
        self._write_reg(self._REG_CTRL_REG1, (0 << 3) | (self._avg & 0x07))
        ctrl2 = self._read_reg(self._REG_CTRL_REG2, 1)[0]
        self._write_reg(self._REG_CTRL_REG2, ctrl2 | 0x01)
        deadline = time.ticks_ms() + 1000 if hasattr(time, 'ticks_ms') else None
        while True:
            status = self._read_reg(self._REG_STATUS, 1)[0]
            if status & self.STATUS_P_DA:
                break
            if deadline is not None and time.ticks_diff(deadline, time.ticks_ms()) <= 0:
                self._write_reg(self._REG_CTRL_REG1, (saved_odr << 3) | (self._avg & 0x07))
                raise OSError('LPS28DFW one-shot timeout (STATUS=0x{:02X})'.format(status))
            time.sleep(0.005)
        result = self.read()
        self._write_reg(self._REG_CTRL_REG1, (saved_odr << 3) | (self._avg & 0x07))
        return result

    def is_data_ready(self):
        """Return True if STATUS.P_DA is set (new pressure sample available).

        Returns:
            bool: True if STATUS bit 0 is asserted.
        """
        return bool(self._read_reg(self._REG_STATUS, 1)[0] & self.STATUS_P_DA)

    def set_offset(self, offset_hpa):
        """Program the one-point calibration offset (RPDS).

        Args:
            offset_hpa: Offset in hPa to subtract from subsequent pressure
                readings (signed). Sensitivity is 4096 LSB/hPa (Mode 1) or
                2048 LSB/hPa (Mode 2).
        """
        if self._fs_mode == 0:
            raw = int(offset_hpa * self._SENSITIVITY_LSB_PER_HPA_MODE1)
        else:
            raw = int(offset_hpa * self._SENSITIVITY_LSB_PER_HPA_MODE2)
        if raw < 0:
            raw += 0x10000
        self._write_reg(0x1A, raw & 0xFF)
        self._write_reg(0x1B, (raw >> 8) & 0xFF)

    def softreset(self):
        """Issue a software reset and wait for the chip to reboot (~2 ms)."""
        self._write_reg(self._REG_CTRL_REG2, 0x02)
        time.sleep(self._BOOT_WAIT_MS / 1000.0)

    def fifo_configure(self, mode=0, wtm=0, stop_on_wtm=False):
        """Configure FIFO mode, watermark level, and stop-on-watermark.

        Args:
            mode: FIFO mode (0=bypass, 1=FIFO, 2=continuous; +4/+5/+6 for
                triggered variants Bypass-to-FIFO / Bypass-to-continuous /
                Continuous-to-FIFO).
            wtm: Watermark level, 0–127 samples.
            stop_on_wtm: True to limit FIFO depth to the watermark.
        """
        if mode == self.FIFO_BYPASS:
            self._write_reg(0x14, 0x00)
        trig = 1 if mode >= 4 else 0
        f_mode = mode & 0x03
        ctrl = (trig << 2) | (int(bool(stop_on_wtm)) << 3) | f_mode
        self._write_reg(0x14, ctrl)
        self._write_reg(0x15, wtm & 0x7F)

    def fifo_read(self, count):
        """Drain up to `count` pressure samples from the FIFO.

        Args:
            count: Number of samples to read.

        Returns:
            list[float]: Pressure values in hPa, most-recent last.
        """
        if count < 0:
            raise ValueError('count must be >= 0')
        if count > 128:
            count = 128
        if count == 0:
            return []
        raw = self._read_reg(0x78, count * 3)
        out = []
        sens = self._SENSITIVITY_LSB_PER_HPA_MODE1 if self._fs_mode == 0 else self._SENSITIVITY_LSB_PER_HPA_MODE2
        for i in range(count):
            base = i * 3
            v = (raw[base + 2] << 16) | (raw[base + 1] << 8) | raw[base]
            if v & 0x800000:
                v -= 0x1000000
            out.append(v / sens)
        return out

    def fifo_level(self):
        """Return the number of unread samples in the FIFO.

        Returns:
            int: 0 (empty) through 128 (full).
        """
        return self._read_reg(0x25, 1)[0]

    def set_threshold(self, threshold_hpa, high=True, low=True):
        """Program the pressure threshold and enable interrupt sources.

        Args:
            threshold_hpa: Pressure threshold in hPa (15-bit unsigned).
                Sensitivity is 16 LSB/hPa (Mode 1) or 8 LSB/hPa (Mode 2).
            high: True to assert PH when pressure exceeds the threshold.
            low: True to assert PL when pressure falls below the threshold.
        """
        if self._fs_mode == 0:
            raw = int(threshold_hpa * 16)
        else:
            raw = int(threshold_hpa * 8)
        if raw < 0:
            raw = 0
        if raw > 0x7FFF:
            raw = 0x7FFF
        self._write_reg(0x0C, raw & 0xFF)
        self._write_reg(0x0D, (raw >> 8) & 0x7F)
        cfg = self._read_reg(self._REG_INTERRUPT_CFG, 1)[0]
        cfg &= ~0x03
        if high:
            cfg |= 0x01
        if low:
            cfg |= 0x02
        self._write_reg(self._REG_INTERRUPT_CFG, cfg)

    def chip_id(self):
        """Read the WHO_AM_I register.

        Returns:
            int: Chip ID; expect 0xB4 for LPS28DFW.
        """
        return self._read_reg(self._REG_WHO_AM_I, 1)[0]